packages/xhr-http-handler/src/xhr-http-handler.ts

```diff
@@ -0,0 +1,222 @@
+ */
+export class XhrHttpHandler extends EventEmitter implements HttpHandler {
+  /**
+   * Events emitted as an EventEmitter.
+   */
+  public static EVENTS = {
+    /**
+     * Emitted for xhr on progress.
+     * Payload is the native ProgressEvent.
+     */
+    PROGRESS: "progress",
+    /**
+     * Emitted for xhr.upload on progress.
+     * Payload is the native ProgressEvent.
+     */
+    UPLOAD_PROGRESS: "upload.progress",
+  };
+
+  private config?: XhrHttpHandlerOptions;
+  private readonly configProvider?: Provider<XhrHttpHandlerOptions>;
+
+  public constructor(options?: XhrHttpHandlerOptions | Provider<XhrHttpHandlerOptions | undefined>) {
+    super();
+    if (typeof options === "function") {
+      this.configProvider = async () => (await options()) || {};
+    } else {
+      this.config = options ?? {};
+    }
+  }
+
+  public destroy(): void {
+    // Do nothing. Connection pooling is handled by the browser.
+  }
+
+  public async handle(
+    request: HttpRequest,
+    { abortSignal }: HttpHandlerOptions = {}
+  ): Promise<{ response: HttpResponse }> {
+    if (!this.config && this.configProvider) {
+      this.config = await this.configProvider();
+    }
+    const requestTimeoutInMs = Number(this.config!.requestTimeout) | 0;
+
+    // if the request was already aborted, prevent doing extra work
+    if (abortSignal?.aborted) {
+      const abortError = new Error("Request aborted");
+      abortError.name = "AbortError";
+      throw abortError;
+    }
+
+    let path: string = request.path;
+    if (request.query) {
+      const queryString = buildQueryString(request.query);
+      if (queryString) {
+        path += `?${queryString}`;
+      }
+    }
+
+    const { port, method } = request;
+    const url = `${request.protocol}//${request.hostname}${port ? `:${port}` : ""}${path}`;
+    const body = method === "GET" || method === "HEAD" ? undefined : request.body;
+
+    const xhr: XMLHttpRequest =
+      typeof this.config!.xhrFactory === "function" ? await this.config!.xhrFactory() : new XMLHttpRequest();
+
+    const raceOfPromises: Promise<{ response: HttpResponse }>[] = [
+      new Promise((resolve, reject) => {
+        xhr.upload.addEventListener("progress", (event: ProgressEvent) => {
+          this.emit(XhrHttpHandler.EVENTS.UPLOAD_PROGRESS, event, request);
+        });
+        xhr.addEventListener("progress", (event: ProgressEvent) => {
+          this.emit(XhrHttpHandler.EVENTS.PROGRESS, event, request);
+        });
+        xhr.addEventListener('error', (error) => {
+          reject(error);
+        })
+        xhr.addEventListener("readystatechange", () => {
+          switch (xhr.readyState) {
+            case XMLHttpRequest.DONE:
+              const body = (() => {
+                switch (xhr.responseType) {
+                  case "document":
+                  case "json":
+                  case "text":
+                  default:
+                    const data = xhr.responseText.split("");
```

Comment:
Are we caching the data and copy here? If so it could limit the handler's usage outside the lib-storage. Anyway we can stream the response data?